experimental/client-csharp/src/Wandb/Run.cs

```diff
@@ -0,0 +1,142 @@
+using System;
+using System.Text;
+using System.Threading.Tasks;
+using WandbInternal;
+using Wandb.Internal;
+
+namespace Wandb
+{
+    public class Run : IDisposable
+    {
+        private readonly SocketInterface _interface;
+        public Settings Settings;
+        public Config Config;
+
+        internal Run(SocketInterface @interface, Settings settings)
+        {
+            _interface = @interface;
+
+            Settings = settings;
+
+            Config = new Config();
+            // Subscribe to ConfigUpdated event
+            Config.ConfigUpdated += OnConfigUpdated;
+        }
+
+        // Callback method
+        private async void OnConfigUpdated(string key, object value)
+        {
+            // Handle the updated configuration
+            await _interface.PublishConfig(key, value); // Example method in SocketInterface
+        }
+
+
+        public async Task Init()
+        {
+            await _interface.InformInit(Settings);
+            Result deliverRunResult = await _interface.DeliverRun(this);
+            if (deliverRunResult.RunResult == null)
+            {
+                throw new Exception("Failed to deliver run");
+            }
+
+            RunUpdateResult runResult = deliverRunResult.RunResult;
+            if (runResult.Error != null)
+            {
+                throw new Exception(runResult.Error.Message);
+            }
+
+            // save project, entity and displa name to settings
+            Settings.Project = runResult.Run.Project;
+            Settings.Entity = runResult.Run.Entity;
+            Settings.DisplayName = runResult.Run.DisplayName;
+            Result result = await _interface.DeliverRunStart(this);
+            if (result.Response == null)
+            {
+                throw new Exception("Failed to deliver run start");
+            }
+            printRunURL();
+        }
+
+
+
+
+        public async Task Log(Dictionary<string, object> data)
```

Comment:
Just to confirm here - we will send line by line and the batching is handled internally?
How can we confirm that all the lines sent have been delivered?

Explicitly:
- recovery from a failed state should be fine - we will simply resend all of them?
- originally, I was considering what would happen if we closed the process, but we've moved past this design now.

Reply:
> Just to confirm here - we will send line by line and the batching is handled internally? 
Correct.

> How can we confirm that all the lines sent have been delivered?
We will block when we call `run.Finish()` until all data is delivered.

> * recovery from a failed state should be fine - we will simply resend all of them?
Correct. This is definitely the case I want thorough tests for, including in the CI.

> * originally, I was considering what would happen if we closed the process, but we've moved past this design now.
Could you clarify? You mean the monitoring process?